<a href="https://colab.research.google.com/github/shawnmuhr/BIOL_398/blob/main/HW_Solutions/biol300_hw3_solns.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Problem 1: Comparing the 18S sequencing of three bird species (20 points)

a) Referencing the pre-class exercise we did for Friday, load in the `bird_18s.txt` file. 

b) Parse the information in this FASTA file, saving `chicken`, `duck`, and `turkey` variables as their corresponding DNA sequences.

c) Write a function with the following function call `SNP_counter(seq1, seq2)` that will return the number of difference (i.e. single nucleotide polymorphisms) between the two sequences inputted.

d) Using the function you wrote in part c, determine the number of difference between these birds' sequences. Display your results with the following print statements:

> `chicken and duck differ by __ basepairs`

> `chicken and turkey differ by __ basepairs`

> `duck and turkey differ by __ basepairs`

e) From your results, what can you conclude about the relative evolutionary relationship between these three birds? Does this make sense to you?


In [ ]:
from google.colab import files
uploaded = files.upload()

Saving 201Syllabus_S22.pdf to 201Syllabus_S22.pdf


In [ ]:
bird_data = uploaded['bird_18S.txt'].decode("utf-8").split('\n')
bird_data

['>chicken',
 'CCAGCTCCAATAGCGTATATTAAAGTTGCTGCAGTTAAAAAGCTCGTAGTTGGATCTTGG',
 'GATCGAGCTGGCGGTCCGCCGCGAGGCGAGCTACCGCCTGTCCCAGCCCCTGTCTCTCGG',
 'CGCCCCCTCGATGCTCTTAACTGAGTGTCCCGCGGGGCCCGAAGCGTTTACTTTGAAAAA',
 'ATTAGAGTGTTCAAAGCAGGCTGGCCGCCGGAATACTCCAGCTAGGAATAATGGAATAGG',
 'ACTCCGGTTCTATTTTGTTGGTTTTCGGAAACGGGGCCATGATTAAGAGGGACGGCCGGG',
 'GGCATTCGTATTGTGCCGCTAGAGGTGAAATTCTTGGACCGGCGCAAGACGAACTAAAGC',
 'GAAA',
 '>duck',
 'CCAGCTCCAATAGCGTATCTTAAAGTTGCTGCAGTTAAAAAGCTCGTAGTTGGATCTTGG',
 'GATGGAGCTGGCGGTCCGCCGCGAGGCGAGCTACCGCCTGTGCCAGCCCCTGCCTGTCGG',
 'CGCCCCCTCGATGCTCTTAAGTGAGTGTCCCGCGGGGCCCGAAGCGTTTACTTTGAAAAA',
 'ATTAGAGTGTTCAAAGCAGGCCGGCCGCCGGAATACTGCAGCTAGGAATAATGGAATAGG',
 'ACTCCGGTTCTATTTTGTTGGTTTTCGGAAACGGGGCCATGATTAAGAGGGACGGCCGGG',
 'GGCATTCGTATTGTGCCGCTAGAGGTGAAATTCTTGGACCGGCGCAAGACGAACTAAAGC',
 'GAAA',
 '>turkey',
 'CCAGCTCCAATAGCGTATCTTAAAGTTGCTGCAGTTAAAAAGCTCGTAGTTGGATCTTGG',
 'GATCGAGCTGGCGGTCCGCCGCGAGGCGAGCTACCGCCTGTCCCAGCCCCTGTCTCTCGG',
 'CGCCCCCTCGATGCTCTTAACTGAGTGTCCCGCGG

First I find the locations of the three headers.

In [ ]:
chicken_loc = bird_data.index(">chicken")
duck_loc = bird_data.index(">duck")
turkey_loc = bird_data.index(">turkey")
print(chicken_loc, duck_loc, turkey_loc)

0 8 16


Using the locations of the headers, I stitch together the intervening sequences with the `join` method. That is, the sequences between the chicken and duck headers correspond to the chicken sequence, while the sequences between the duck and turkey headers correspond to the duck sequence.

In [ ]:
chicken = "".join(bird_data[chicken_loc+1 : duck_loc])
duck = "".join(bird_data[duck_loc+1 : turkey_loc])
turkey = "".join(bird_data[turkey_loc+1:])

Printing out these sequences, they look good! Now we can move on to comparing them! 

In [ ]:
print(chicken)
print(duck)
print(turkey)

CCAGCTCCAATAGCGTATATTAAAGTTGCTGCAGTTAAAAAGCTCGTAGTTGGATCTTGGGATCGAGCTGGCGGTCCGCCGCGAGGCGAGCTACCGCCTGTCCCAGCCCCTGTCTCTCGGCGCCCCCTCGATGCTCTTAACTGAGTGTCCCGCGGGGCCCGAAGCGTTTACTTTGAAAAAATTAGAGTGTTCAAAGCAGGCTGGCCGCCGGAATACTCCAGCTAGGAATAATGGAATAGGACTCCGGTTCTATTTTGTTGGTTTTCGGAAACGGGGCCATGATTAAGAGGGACGGCCGGGGGCATTCGTATTGTGCCGCTAGAGGTGAAATTCTTGGACCGGCGCAAGACGAACTAAAGCGAAA
CCAGCTCCAATAGCGTATCTTAAAGTTGCTGCAGTTAAAAAGCTCGTAGTTGGATCTTGGGATGGAGCTGGCGGTCCGCCGCGAGGCGAGCTACCGCCTGTGCCAGCCCCTGCCTGTCGGCGCCCCCTCGATGCTCTTAAGTGAGTGTCCCGCGGGGCCCGAAGCGTTTACTTTGAAAAAATTAGAGTGTTCAAAGCAGGCCGGCCGCCGGAATACTGCAGCTAGGAATAATGGAATAGGACTCCGGTTCTATTTTGTTGGTTTTCGGAAACGGGGCCATGATTAAGAGGGACGGCCGGGGGCATTCGTATTGTGCCGCTAGAGGTGAAATTCTTGGACCGGCGCAAGACGAACTAAAGCGAAA
CCAGCTCCAATAGCGTATCTTAAAGTTGCTGCAGTTAAAAAGCTCGTAGTTGGATCTTGGGATCGAGCTGGCGGTCCGCCGCGAGGCGAGCTACCGCCTGTCCCAGCCCCTGTCTCTCGGCGCCCCCTCGATGCTCTTAACTGAGTGTCCCGCGGGGCCCGAAGCGTTTACTTTGAAAAAATTAGAGTGTTCAAAGCAGGCCGGCCGCCGGAATACTCCAGCTAGGAATAATGGAATAGGACTCCGGTTCTATTTTGTTGGTTTTCGGAA

In [ ]:
def SNP_counter(seq1, seq2):
  """ Returns the number of differences between two sequences"""

  diff_counter = 0

  # loop through the length of the sequence
  for i in range(len(seq1)):

    # if we find a mismatch, we increment the counter
    if seq1[i] != seq2[i]:
      diff_counter +=1

  return diff_counter

In [ ]:
chicken_v_duck = SNP_counter(chicken, duck)
chicken_v_turkey = SNP_counter(chicken, turkey)
duck_v_turkey = SNP_counter(duck, turkey)

print("chicken and duck differ by", chicken_v_duck, "basepairs")
print("chicken and turkey differ by", chicken_v_turkey, "basepairs")
print("duck and turkey differ by", duck_v_turkey, "basepairs")

chicken and duck differ by 8 basepairs
chicken and turkey differ by 2 basepairs
duck and turkey differ by 6 basepairs


From this, we see that chicken and turkey are most related! We may have expected this as duck is an aquatic bird with webbed feet, etc. But it is satisfying to confirm it in this way. 

# Problem 2: Using GC content to find the pathogenticy island in *Salmonella*

We've spent some time in this class computing the GC content of various sequences, but we will now apply that skill to address a question about *Salmonella*: where does it's pathogenicity arise from? GC content can vary across species, so regions of the *Salmonella* genome that differ greatly in GC content tend to suggest regions that arose from Horizontal Gene Transfer (HGT), meaning it picked up the gene from another organism. Such swapping of genes is very common in bacteria, and it is suspected that the genes that make *Salmonella* pathogenic arose from HGT, meaning those genes will have a different GC content from the rest of the genome. The following exercises will walk you through the process of uncovering this 'pathogenicity island' yourself!

## Problem 2.1 (10 points)

This time, we are interested not just of the GC content of an *entire* sequence, but of chunks along a genome, allowing us to find regions of lower or higher GC content. Write a function that divides a sequence into blocks and computes the GC content for each block, returning a list. The function signature should look like

`GC_blocks(seq, block_size)`

To be clear, if `seq = 'ATGACTACGT'` and `block_size = 4`, the blocks to be considered are

> `ATGA` and `CTAC`

and the function should return `[0.25, 0.5]`. Note that the blocks are non-overlapping and that we don't bother with the end of the sequence that does not fit completely in a block. **Note: Make sure to call your `GC_content` function from before!**

In [ ]:
def GC_content(DNA_seq):
  """Computes the GC content (as a float) given the provided DNA sequence 
  (as a string)"""

  # count Cs and Gs, respectively
  C_count = DNA_seq.count('C')
  G_count = DNA_seq.count('G')

  # comupte GC content as a fraction 
  GC_frac = (C_count + G_count) / len(DNA_seq)

  return GC_frac

In [ ]:
def GC_blocks(seq, block_size):
  """ Returns a list of GC content across the length of the provided 
  sequence, moving along in 'chunks' of length block_size"""

  # figure out how many blocks there will be, rounding down
  blocks = int(len(seq)/block_size)

  GC_blocks = []

  # loop through the number of block
  for i in range(blocks):
    
    # grab the DNA sequence block we are interested in
    block = seq[block_size*i : block_size*(i+1)]

    # compute GC content of this block, and add to our list
    GC_block = GC_content(block)
    GC_blocks.append(GC_block)

  return GC_blocks

Let's verify that it returns the expected result:

In [ ]:
GC_blocks('ATGACTACGT', 4)

[0.25, 0.5]

## Problem 2.2 (15 points)

Write a function that takes as input a sequence, block size, and a threshold GC content, and returns a visualization of the sequence, where `'*'` designates a block that is below the threshold, and `'_'` designates a block that is above the threshold. You would call the function like this: For example,

> `GC_map('ATGACTACGT', 4, 0.4)`

returns `'*_'`, indicating the first block has a GC content below 0.4, while the second block has a GC content above 0.4. Note that bases not included in GC blocks are truncated from the visualization.

In [ ]:
def GC_map(seq, block_size, gc_thresh):

  # call our previous function to get our GC content values
  GC_values = GC_blocks(seq, block_size)

  output_seq = ""

  # loop through GC content values
  for GC_val in GC_values:

    # update the output string based on value of GC content
    if GC_val < gc_thresh:
      output_seq += "*"
    else:
      output_seq += "_"

  return output_seq

Again, let's check that this works as expected:

In [ ]:
GC_map('ATGACTACGT', 4, 0.4)

'*_'

# Problem 2.3 (10 points)

Let's take your GC_map function for a spin, trying it on a portion of the Salmonella genome! First, load in the 'salmonella_spi1_region.txt' file and store the sequence as a single string with no gaps.

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving salmonella_spi1_region.txt to salmonella_spi1_region.txt


Let's look at the first ten lines to get a sense of the structure.

In [ ]:
salmonella = uploaded['salmonella_spi1_region.txt'].decode("utf-8").split('\n')
salmonella[:10]

['>gi|821161554|gb|CP011428.1| Salmonella enterica subsp. enterica strain YU39, complete genome, subsequence 3000000 to 3200000',
 'AAAACCTTAGTAACTGGACTGCTGGGATTTTTCAGCCTGGATACGCTGGTAGATCTCTTC',
 'ACGATGGACAGAAACTTCTTTCGGGGCGTTCACGCCAATACGCACCTGGTTGCCCTTCAC',
 'CCCTAAAACTGTCACGGTGACCTCATCGCCAATCATGAGGGTCTCACCAACTCGACGAGT',
 'CAGAATCAGCATTCTTTGCTCCTTGAAAGATTAAAAGAGTCGGGTCTCTCTGTATCCCGG',
 'CATTATCCATCATATAACGCCAAAAAGTAAGCGATGACAAACACCTTAGGTGTAAGCAGT',
 'CATGGCATTACATTCTGTTAAACCTAAGTTTAGCCGATATACAAAACTTCAACCTGACTT',
 'TATCGTTGTCGATAGCGTTGACGTAAACGCCGCAGCACGGGCTGCGGCGCCAACGAACGC',
 'TTATAATTATTGCAATTTTGCGCTGACCCAGCCTTGTACACTGGCTAACGCTGCAGGCAG',
 'AGCTGCCGCATCCGTACCACCGGCTTGCGCCATGTCCGGACGACCGCCACCCTTACCGCC']

Since this file only has one sequence, we can use the `join` method again, this time on *all* the lines in the file (excluding the header). We do that below, and take a look at the first 1000 characters to make sure it looks correct.

In [ ]:
salmonella_seq = "".join(salmonella[1:])
salmonella_seq[:1000]

'AAAACCTTAGTAACTGGACTGCTGGGATTTTTCAGCCTGGATACGCTGGTAGATCTCTTCACGATGGACAGAAACTTCTTTCGGGGCGTTCACGCCAATACGCACCTGGTTGCCCTTCACCCCTAAAACTGTCACGGTGACCTCATCGCCAATCATGAGGGTCTCACCAACTCGACGAGTCAGAATCAGCATTCTTTGCTCCTTGAAAGATTAAAAGAGTCGGGTCTCTCTGTATCCCGGCATTATCCATCATATAACGCCAAAAAGTAAGCGATGACAAACACCTTAGGTGTAAGCAGTCATGGCATTACATTCTGTTAAACCTAAGTTTAGCCGATATACAAAACTTCAACCTGACTTTATCGTTGTCGATAGCGTTGACGTAAACGCCGCAGCACGGGCTGCGGCGCCAACGAACGCTTATAATTATTGCAATTTTGCGCTGACCCAGCCTTGTACACTGGCTAACGCTGCAGGCAGAGCTGCCGCATCCGTACCACCGGCTTGCGCCATGTCCGGACGACCGCCACCCTTACCGCCCACCTGCTGAGCGACCATCCCAATCAATTCCCCTGCTTTAACCCGGTCGGTCACATCCTTCGACACGCCCGCAATCAGAGAAACCTTACCTTCAACAACCGTTGCCAGTACGATAACGGTAGACCCCAGTTGATTTTTCAGATCATCAACCATGGTTCGCAGCATTTTCGGCTCAATACCAGCAAGCTCGCTAACCAGGAGCTTCACGCCGTTGAGATCGACCGCTTTACTGGAAAGATTTGCACTCTCCTGCGCTGCGGCCTGGTCCTTCAACTGCTGCAACTCTTTTTCCAGCTGACGTGTACGTTCCAGCACGGCACGCACTTTGTCGCCCAGATTCTGGCTGTCGCCCTTCAGCAGATGCGCAATATCGTTTAAGCGATCGCTTTGCGCATGAACGGTGGCCATAGCGCCTTCACCGGTTACCGCCTCAATACGACGAATGCCCGCTGCGGTGCC

## Problem 2.4 (5 points)

Run your `GC_map()` function on the salmonella sequence with `block_size = 1000` and  `gc_thresh = 0.45`. 

Looking at your visualization of GC content, about how far through the sequence (e.g. 10%, 80%) do you notice a region of lower GC content blocks? This is the pathogenicity island! This region is suspected to be where the genes that make *Salmonella* pathogenic lie!

Finally we are ready to run our `GC_map` function on the section of the salmonella genome!

In [ ]:
GC_map(salmonella_seq, 1000, 0.45)

'_______________________________________****___***_*_**______**_____*_*___***_**_____________________________________*_____________*________________________*____________________________________________'

It looks like there is a region spanning 20 to 40% of the way through this sequence that has many low GC-content blocks, indicated by the `*`s. There's also a few spurious chunks further down, but that larger region is likely the pathogenicity island. 

# Problem 3: Exploring a PDB file (30 points)

For this last problem, we will explore a new type of data file, a protein database, or PDB file. We will be looking at SARS-COV2 spike protein [found here on PDB website](https://www.rcsb.org/structure/6XLU), although you will be manipualting the .txt file for this problem.

You may wish to open the file to get a sense of its structure. If you scroll down far enough, you will see a bunch of lines that begin with `ATOM`. These lines are the ones that describe the structure of the protein, by specifying the precise x-, y-, and z-coordinates of each atom in the structure. Each line contains information like atom number, atom identity (carbon, nitrogen, etc.), which amino acid the atom belongs to, and which subunit chain the amino acid belongs to. As an example, I show the first line below:
> `ATOM   1     N N   . ALA A  1 14   ? 137.193 160.456 208.534 1.00 62.75 ? 27   ALA A N   1`
 
See the [PDB formatting guidelines](http://www.wwpdb.org/documentation/file-format-content/format33/sect9.html#ATOM) for a full description of what each column of information contains.

a) Read in the `'6xlu.txt'` file. 

b) Parse the information from the file to only get the ATOMs found in chain `'A'`. Note the chainID can be found in the 25th index of a given line in this file. Save each ATOM line corresponding to chain `'A'` to a list.

c) Display the first 10 and last 10 lines of your list to make sure you've isolated only the atoms we care about.

d) How many atoms are in chain `'A'`?

e) How many amino acids are in chain `'A'`?

f) Modify your code to repeat the process for chains `'B'` and `'C'`. How many amino acids do they have? What do you suspect this implies for the overall structure of the spike protein? 

In [ ]:
from google.colab import files
uploaded = files.upload()

Saving 6xlu.txt to 6xlu.txt


First, let's read in the file. I will split by the new line character, to get a list of each line.

In [ ]:
spike = uploaded['6xlu.txt'].decode("utf-8").split('\n')
spike[:10]

['data_6XLU',
 '# ',
 '_entry.id   6XLU ',
 '# ',
 '_audit_conform.dict_name       mmcif_pdbx.dic ',
 '_audit_conform.dict_version    5.352 ',
 '_audit_conform.dict_location   http://mmcif.pdb.org/dictionaries/ascii/mmcif_pdbx.dic ',
 '# ',
 'loop_',
 '_database_2.database_id ']

While a function wasn't explicitly asked for, I see that I will need to run the same code for chains A, B, and C. So I write a function to grab only the atoms from a given chain.

In [ ]:
def grab_chain(pdb_lines, chain_name):
  """Returns the lines the corresponds to desired chain.
  The inputted pdb_lines must already be parsed to be a list
  of the file lines"""

  chain = []

  # loop through the lines of the pdb file
  for line in pdb_lines:

    # if line starts with ATOM and has the desired chain, 
    # we add it to our list
    if line[0:4] == "ATOM":
      if line[25] == chain_name:
        chain.append(line)
    
  return chain

Let's run this function on chains A, B, and C in turn. The length of the resulting chain specifies how many atoms are in the given chain. We see that they have a similar, but not exactly the same, number of atoms.

In [ ]:
chain_A = grab_chain(spike, "A")
chain_B = grab_chain(spike, "B")
chain_C = grab_chain(spike, "C")
print("Chain A contains", len(chain_A), "atoms.")
print("Chain B contains", len(chain_B), "atoms.")
print("Chain C contains", len(chain_C), "atoms.")

Chain A contains 8276 atoms.
Chain B contains 8302 atoms.
Chain C contains 8294 atoms.


Let's display  the first and last 10 lines of chain_A:

In [ ]:
chain_A[:10]

['ATOM   1     N N   . ALA A  1 14   ? 137.193 160.456 208.534 1.00 62.75 ? 27   ALA A N   1 ',
 'ATOM   2     C CA  . ALA A  1 14   ? 136.846 161.310 207.402 1.00 62.75 ? 27   ALA A CA  1 ',
 'ATOM   3     C C   . ALA A  1 14   ? 138.012 162.215 207.025 1.00 62.75 ? 27   ALA A C   1 ',
 'ATOM   4     O O   . ALA A  1 14   ? 138.903 162.453 207.837 1.00 62.75 ? 27   ALA A O   1 ',
 'ATOM   5     C CB  . ALA A  1 14   ? 135.613 162.138 207.720 1.00 62.75 ? 27   ALA A CB  1 ',
 'ATOM   6     N N   . TYR A  1 15   ? 137.999 162.715 205.789 1.00 61.90 ? 28   TYR A N   1 ',
 'ATOM   7     C CA  . TYR A  1 15   ? 139.055 163.573 205.274 1.00 61.90 ? 28   TYR A CA  1 ',
 'ATOM   8     C C   . TYR A  1 15   ? 138.468 164.515 204.239 1.00 61.90 ? 28   TYR A C   1 ',
 'ATOM   9     O O   . TYR A  1 15   ? 137.430 164.231 203.641 1.00 61.90 ? 28   TYR A O   1 ',
 'ATOM   10    C CB  . TYR A  1 15   ? 140.183 162.766 204.643 1.00 61.90 ? 28   TYR A CB  1 ']

In [ ]:
chain_A[-10:]

['ATOM   8267  C CB  . ASP A  1 1133 ? 182.881 182.210 97.826  1.00 66.32 ? 1146 ASP A CB  1 ',
 'ATOM   8268  C CG  . ASP A  1 1133 ? 183.120 181.238 98.977  1.00 66.32 ? 1146 ASP A CG  1 ',
 'ATOM   8269  O OD1 . ASP A  1 1133 ? 184.035 181.493 99.783  1.00 66.32 ? 1146 ASP A OD1 1 ',
 'ATOM   8270  O OD2 . ASP A  1 1133 ? 182.412 180.214 99.071  1.00 66.32 ? 1146 ASP A OD2 1 ',
 'ATOM   8271  N N   . SER A  1 1134 ? 183.248 185.826 97.084  1.00 63.52 ? 1147 SER A N   1 ',
 'ATOM   8272  C CA  . SER A  1 1134 ? 183.507 186.843 96.069  1.00 63.52 ? 1147 SER A CA  1 ',
 'ATOM   8273  C C   . SER A  1 1134 ? 184.900 186.772 95.439  1.00 63.52 ? 1147 SER A C   1 ',
 'ATOM   8274  O O   . SER A  1 1134 ? 185.111 187.250 94.325  1.00 63.52 ? 1147 SER A O   1 ',
 'ATOM   8275  C CB  . SER A  1 1134 ? 183.318 188.225 96.691  1.00 63.52 ? 1147 SER A CB  1 ',
 'ATOM   8276  O OG  . SER A  1 1134 ? 181.985 188.404 97.121  1.00 63.52 ? 1147 SER A OG  1 ']

From this output, we can confirm that there are 8276 atoms in chain A, from the number that appears immediately after `ATOM`. We can also see that two columns over from the chain name 'A', there is a specification for the amino acid number. That is, all the atoms corresponding to ALA are in the 14th amino acid (interesting to think what happened to the first 13!), while all the atoms corresponding to SER are in the 1134th amino acid in the chain.

For the purposes of this exercise, I just wanted you to get familiar with the structure of this file, so it was sufficient to assume that the chain goes from 14 to 1134, or is 1121 amino acids long. (Below, I will delve more deeply into this).

To see the length of chain B and C in a similar way, let's just print out the first and last line of each chain and similarly look at that numeric value for the amino acid. 

In [ ]:
print(chain_B[0])
print(chain_B[-1])
print(chain_C[0])
print(chain_C[-1])

ATOM   8277  N N   . ALA B  1 14   ? 234.997 157.197 208.500 1.00 60.77 ? 27   ALA B N   1 
ATOM   16578 O OG  . SER B  1 1134 ? 186.570 183.701 97.382  1.00 62.19 ? 1147 SER B OG  1 
ATOM   16579 N N   . ALA C  1 14   ? 188.766 243.345 208.464 1.00 60.93 ? 27   ALA C N   1 
ATOM   24872 O OG  . SER C  1 1134 ? 190.458 190.251 97.144  1.00 66.43 ? 1147 SER C OG  1 


Interesting! All three chains both start and end with same number of amino acids, suggesting that the spike protein is what we would call a homo-trimer, or a protein made up of three identical (or nearly identical) chains!

## Extension

Everything below is beyond what I expected for this problem, but devles more deeply into the number of amino acids. That is, I assumed that the number incremented for each chain from 14 to 1134, but I would like to verify if that is the case.

From a little careful counting, I find that the amino acid number can be found at characters 30 to 33 in each line. I will use this information to parse out if the numbers do truly increment.

In [ ]:
def missing_aa(chain):
  """returns the indices of amino acids that are missing from a given chain"""

  # find the first and last amino acid number in the chain
  start = int(chain[0][30:34])
  end = int(chain[-1][30:34])

  # keep a list of numbers found
  aa_nums = []

  # loop through chain and get the aa number at each atom
  for atom in chain:
    aa_num = int(atom[30:34])
    aa_nums.append(aa_num)

  # keep a list of missing amino acid number
  missing_aa = []

  # loop from the start and end and find which numbers are missing
  for i in range(start,end+1):
    if i not in aa_nums:
      missing_aa.append(i)

  return missing_aa

In [ ]:
print(missing_aa(chain_A))
print(missing_aa(chain_B))
print(missing_aa(chain_C))

[57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 238, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 432, 433, 664, 665, 666, 667, 668, 669, 670, 671, 672, 673, 674, 675, 816, 817, 818, 819, 820]
[57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 251, 432, 433, 664, 665, 666, 667, 668, 669, 670, 671, 672, 673, 674, 675]
[57, 58, 59, 60, 61, 62, 63, 64, 65, 66, 131, 132, 133, 134, 135, 136, 137, 138, 139, 140, 160, 161, 162, 163, 164, 165, 166, 167, 168, 169, 170, 171, 172, 239, 240, 241, 242, 243, 244, 245, 246, 247, 248, 249, 250, 432, 433, 664, 665, 666, 667, 668, 669, 670, 671, 672, 673, 674, 675, 816, 817]


We see that each chain is in fact missing some amino acid indices! This suggests that there is some post-translational processing that is happening, clipping out a few amino acids at various points along the protein. 

To get the true length of each chain, we can subract the number of missing indicies from the expected length of 1120.

In [ ]:
print("Chain A is", 1121 - len(missing_aa(chain_A)), "amino acids long.")
print("Chain B is", 1121 - len(missing_aa(chain_B)), "amino acids long.")
print("Chain C is", 1121 - len(missing_aa(chain_C)), "amino acids long.")

Chain A is 1057 amino acids long.
Chain B is 1061 amino acids long.
Chain C is 1060 amino acids long.


We see they are nearly the same length, just with slight differences in which specific amino acids were clipped out.